File: day11/day11b.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

// Returns whether there exists an occupied seat in the given direction from seats[row][col]
bool occupied_in_direction(const std::vector<std::string> &seats, int row, int col,
                           int row_stride, int col_stride) {
  int seek_row = row + row_stride;
  int seek_col = col + col_stride;

  while(0 <= seek_row && seek_row < seats.size() && 0 <= seek_col && seek_col < seats[seek_row].length()) {
    if(seats[seek_row][seek_col] == '#')
      return true;
    else if(seats[seek_row][seek_col] == 'L')
      return false;
    seek_row += row_stride;
    seek_col += col_stride;
  }
  return false;
}

// Consider the first visible seat in all eight directions
int count_neighbors(const std::vector<std::string> &seats, int row, int col) {
  int num_neighbors = 0;
  for(int row_stride = -1; row_stride <= 1; row_stride++) {
    for(int col_stride = -1; col_stride <= 1; col_stride++) {
      if(row_stride != 0 || col_stride != 0) {
        if(occupied_in_direction(seats, row, col, row_stride, col_stride)) {
          num_neighbors++;
        }
      }
    }
  }
  return num_neighbors;
}

bool check_equal(const std::vector<std::string> &seats, const std::vector<std::string> &new_seats) {
  if(seats.size() != new_seats.size()) {
    return false;
  }

  for(int row = 0; row < seats.size(); row++) {
    if(seats[row] != new_seats[row]) {
      return false;
    }
  }
  return true;
}

int main() {
  // Read input from file
  std::ifstream file("day11.txt");
  std::string line;
  std::vector<std::string> seats;

  while(std::getline(file, line)) {
    seats.push_back(line);
  }

  // Keep applying rules until seats no longer change
  std::vector<std::string> new_seats(seats);
  while(true) {
    // Apply rules
    for(int row = 0; row < seats.size(); row++) {
      for(int col = 0; col < seats[row].length(); col++) {
        if(seats[row][col] != '.') {
          int neighbors = count_neighbors(seats, row, col);
          if(neighbors == 0) {
            // Empty seat that sees no occupied seats becomes occupied
            new_seats[row][col] = '#';
          } else if(neighbors >= 5) {
            // Occupied seat that sees >=5 occupied seats becomes empty
            new_seats[row][col] = 'L';
          } else {
            // Otherwise, seat state does not change
            new_seats[row][col] = seats[row][col];
          }
        }
      }
    }
    
    // Check whether seat pattern has changed, and return result if no change
    if(check_equal(seats, new_seats)) {
      int occupant_count = 0;
      for(int row = 0; row < seats.size(); row++) {
        for(int col = 0; col < seats[row].length(); col++) {
          if(seats[row][col] == '#') {
            occupant_count++;
          }
        }
      }
      std::cout << occupant_count << std::endl;
      return 0;
    }

    seats.swap(new_seats);
  }

  return 0;
}

File: day11/day11b.py
#!/usr/bin/python3

def check_equal(seats1, seats2):
    num_rows = len(seats1)
    num_cols = len(seats1[0])
    if num_rows != len(seats2):
        return False
    if num_cols != len(seats2[0]):
        return False
    for row in range(num_rows):
        for col in range(num_cols):
            if seats1[row][col] != seats2[row][col]:
                return False
    return True

# Returns whether there exists an occupied seat in the given direction from seats[row][col]
def occupied_in_direction(seats, row, col, row_stride, col_stride):
    num_rows = len(seats)
    num_cols = len(seats[0])
    seek_row = row + row_stride
    seek_col = col + col_stride

    while (0 <= seek_row) and (seek_row < num_rows) and (0 <= seek_col) and (seek_col < num_cols):
        if seats[seek_row][seek_col] == '#':
            return True
        elif seats[seek_row][seek_col] == 'L':
            return False
        seek_row += row_stride
        seek_col += col_stride
    return False

# Consider the first visible seat in all eight directions
def count_neighbors(seats, row, col):
    num_neighbors = 0
    for row_stride in [-1,0,1]:
        for col_stride in [-1,0,1]:
            if row_stride != 0 or col_stride != 0:
                num_neighbors += occupied_in_direction(seats, row, col, row_stride, col_stride)
    return num_neighbors
    
def apply_rules(seats):
    new_seats = [list(row) for row in seats]
    for row in range(len(seats)):
        for col in range(len(seats[0])):
            if seats[row][col] != '.':
                neighbors = count_neighbors(seats, row, col)
                if neighbors == 0:
                    # Empty seats that see no occupied seats become occupied
                    new_seats[row][col] = '#'
                elif neighbors >= 5:
                    # Occupied seats that see >=5 occupied seats become empty
                    new_seats[row][col] = 'L'
                else:
                    # Otherwise, seat state doesn't change
                    new_seats[row][col] = seats[row][col]
    return new_seats

def count_occupants(seats):
    return sum((sum(1 for char in row if char == '#')) for row in seats)

with open('day11.txt', 'r') as f:
    lines = f.read().strip().split('\n')
    seats = [list(line) for line in lines]

    # Keep applying rules until seats no longer change
    while True:
        new_seats = apply_rules(seats)
        if check_equal(seats, new_seats):
            seats = new_seats
            break
        else:
            seats = new_seats

    print(count_occupants(seats))
